### PR TITLE
Improved SQL management

### PR DESCRIPTION
**Pull request by rowanseymour**
Fully aware this likely requires a bit more discussion but hopefully this PR is a good starting point for that.

Current process for modifying database objects in migrations:

1. Create new app specific sql state file based on last one for that app
2. Make changes in SQL file
3. Make copy of state file as `current_<app>.sql`
3. Create migration to run new SQL state file

Downsides of this are 1) need to manually duplicate the state file and 2) need to run entire state file each time which isn't scaling well.

New process would be:

1. Lookup object to be modified in `current_<type>.sql` and use that as starting point
2. Add the CREATE / DROP operation directly in new migration with `RunSQL`
3. Run `/.manage.py collect_sql` to generate new SQL state files.

In this process, SQL state files here are only used for enabling easier code reviewing and for initial state if we do another reset of migrations.

### CODE REVIEW

File: temba/utils/management/commands/collect_sql.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        if by_type[SqlType.FUNCTION]:
+            self.write_dump('functions', by_type[SqlType.FUNCTION], output_dir)
+        if by_type[SqlType.TRIGGER]:
+            self.write_dump('triggers', by_type[SqlType.TRIGGER], output_dir)
+
+    def write_dump(self, type_label, operations, output_dir):
+        filename = '%s/current_%s.sql' % (output_dir, type_label)
+
+        with open(filename, 'w') as f:
+            header = '-- Generated by collect_sql on %s\n\n' % now().strftime("%Y-%m-%d %H:%M")
```
Comment (norkans7) [inline temba/utils/management/commands/collect_sql.py]: use a naive datetime

**Comment by norkans7** [comment]
what if we have backward migrations?

File: temba/utils/management/commands/collect_sql.py (hunk 1 above)
Comment (rowanseymour) [inline temba/utils/management/commands/collect_sql.py]: Updated

**Comment by rowanseymour** [comment]
@norkans7 this command generates sql which matches the state produced by migrations, so as long as our migrations are correct, the sql will be correct.

**Comment by rowanseymour** [comment]
@ericnewcomer there is - it's just the diff is too big to display on the changes tab

When the last IVR migrations are merged and I'll update this and off we go. We could move the old sql files so it's clear we're not doing that anymore - or embed their sql back into the original python migrations?

**Comment by ericnewcomer** [comment]
Ah, see it now.

As for the migration-specific sql files, no reason we can't still continue with that model. The big difference here is that we just put the new stuff in those files instead of grabbing all the triggers / functions for an app. I still much prefer having sql in sql files instead of quoted in migrations.

**Comment by rowanseymour** [comment]
Maybe it depends on quantity of SQL? Separate SQL files would get a bit annoying for one line indexes

**Comment by ericnewcomer** [comment]
Sure, just saying we need'nt _abandon_ that approach. The decision here is that we install deltas instead of the full state of the world. If that delta is a lot of sql, then use a separate file and InstallSQL just as we have been doing.

**Comment by ericnewcomer** [comment]
Actually, @nicpottier just reminded me that pycharm can syntax highlight embedded sql in a big ass string. So, I'm fine ditching external files and keeping things in one file.

**Comment by rowanseymour** [comment]
I've made change so that by default, operations are sorted by object name. This makes for much nicer diffs because otherwise if you modify an object, it jumps to the bottom of the dump.

For an example of a nice diff see https://github.com/nyaruka/rapidpro/compare/make_sql...nyaruka:more_better_squashing?expand=1

If we ever need dumps with the original order (maybe if we were resetting all sql again, and functions have to be declared in order?), just regenerate with `./manage.py collect_sql --preserve-order`

**Pull request merged**